File: Dataset/cpp/FileToNumbers.cpp

//  FileToNumbers.cpp  (c) Kari Laitinen

//  http://www.naturalprogramming.com

//  2006-06-09  File created.
//  2006-06-09  Last modification.

//  This is the C++ version of corresponding Java/C#/Python
//  programs. This program is not presented in the C++ book.

#include  <iostream.h>
#include  <fstream.h>  // C++ classes for file handling.
#include  <iomanip.h>
#include  <sstream>    // Class stringstream
#include  <string>     // Class string

int main( int   number_of_command_line_arguments,
          char* command_line_arguments[] )
{
   if ( number_of_command_line_arguments  !=  2 )
   {
      cout << "\n You have to command this program as: \n"
              "\n  FileToNumbers file.ext \n" ;
   }
   else
   {
      ifstream  file_to_read( command_line_arguments[ 1 ],
                               ios::binary ) ;

      if ( file_to_read.fail() )
      {
         cout << "\n Cannot open " << command_line_arguments[ 1 ] ;
      }
      else
      {
         while ( ! file_to_read.eof() )
         {
            unsigned char  bytes_from_file[ 16 ] ;

            file_to_read.read( bytes_from_file,  16 ) ;

            int number_of_bytes_read  =  file_to_read.gcount() ;

            stringstream  stream_of_bytes  ;
            string        bytes_as_characters  =  "" ;

            for ( int byte_index  =  0 ;
                      byte_index  <  number_of_bytes_read ;
                      byte_index  ++ )
            {
               stream_of_bytes <<  " "  <<  setfill( '0' )  <<  hex
                               <<  setw( 2 )
                               <<  (int) bytes_from_file[ byte_index ]  ;

               if ( bytes_from_file[ byte_index ]  >=  ' ' )
               {
                  bytes_as_characters.append( 1,
                                              bytes_from_file[ byte_index ] );
               }
               else
               {
                  bytes_as_characters.append( 1, ' ' ) ;
               }
            }

            string  line_of_bytes ;

            getline( stream_of_bytes, line_of_bytes ) ;

            cout <<  "\n"  <<  left  <<  setw( 48 ) <<  line_of_bytes
                 <<  "  "  <<  bytes_as_characters ;
         }
      }
   }
}

File: Dataset/cs/FileToNumbers.cs

//  FileToNumbers.cs  (c) 2003 Kari Laitinen

//  www.naturalprogramming.com

using System ;
using System.IO ;    // Classes for file handling.
using System.Text ;  // Class StringBuilder etc.

class FileToNumbers
{
   static void Main( string[] command_line_parameters )
   {
      if ( command_line_parameters.Length  ==  1 )
      {
         try
         {
            FileStream file_to_read  =
                           File.OpenRead( command_line_parameters[ 0 ] ) ;

            byte[]  bytes_from_file  =  new  byte[ 16 ] ;

            int number_of_bytes_read  ;

            while (( number_of_bytes_read  =
                       file_to_read.Read( bytes_from_file, 0,
                                           bytes_from_file.Length ))  >  0 )
            {
               StringBuilder  line_of_bytes  =  new  StringBuilder() ;
               StringBuilder  bytes_as_characters  =  new  StringBuilder () ;

               for ( int byte_index  =  0 ;
                         byte_index  <  number_of_bytes_read ;
                         byte_index  ++ )
               {
                  line_of_bytes.Append( " "
                        +  bytes_from_file[ byte_index].ToString( "X2" )  ) ;

                  char  byte_as_character  =
                                  (char) bytes_from_file[ byte_index ] ;

                  if ( byte_as_character  >=  ' ' )
                  {
                     bytes_as_characters.Append( byte_as_character ) ;
                  }
                  else
                  {
                     bytes_as_characters.Append( ' ' ) ;
                  }
               }

               Console.Write( "\n" +  line_of_bytes.ToString().PadRight(16 * 3 )
                           +  "  "  +  bytes_as_characters.ToString() ) ;
            }

            file_to_read.Close() ;
         }
         catch ( FileNotFoundException )
         {
            Console.Write( "\n Cannot open file "
                        +  command_line_parameters[ 0 ] ) ;
         }
      }
      else
      {
         Console.Write( "\n You have to command this program as: \n"
                     +  "\n  FileToNumbers file.ext \n") ;
      }
   }
}




